### PR TITLE
Add etcd lock

### PR DESCRIPTION
**Pull request by totalo**
Add etcd lock

### CODE REVIEW

**Comment by tristaZero** [comment]
Hi is there any issue relevant to this PR? @totalo

**Comment by totalo** [comment]
> Hi is there any issue relevant to this PR? @totalo

No, I just found todo and tried to fix it.:-D

**Comment by tristaZero** [comment]
Aha, @totalo Thanks, @menghaoranss Please have a look here.

**Comment by menghaoranss** [comment]
> Aha, @totalo Thanks, @menghaoranss Please have a look here.

ok.

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java [hunk 1]
```diff
@@ -129,23 +129,36 @@ private Type getEventChangedType(final WatchEvent event) {
...
-        // TODO
-        return false;
+        try {
+            long leaseId = client.getLeaseClient().grant(etcdProperties.getValue(EtcdPropertyKey.TIME_TO_LIVE_SECONDS)).get().getID();
+            client.getLockClient().lock(ByteSequence.from(key, StandardCharsets.UTF_8), leaseId).get(time, unit);
+            return true;
+            // CHECKSTYLE:OFF
+        } catch (final Exception ex) {
+            // CHECKSTYLE:ON
+            return false;
```
Comment (menghaoranss) [inline shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java:142]: The exception need be printed by @slf4j

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java [hunk 2]
```diff
@@ -129,23 +129,36 @@ private Type getEventChangedType(final WatchEvent event) {
...
+
     @Override
     public void releaseLock(final String key) {
-        // TODO
+        try {
+            client.getLockClient().unlock(ByteSequence.from(key, StandardCharsets.UTF_8)).get(etcdProperties.getValue(EtcdPropertyKey.CONNECTION_TIMEOUT_SECONDS), TimeUnit.SECONDS);
+            // CHECKSTYLE:OFF
+        } catch (final Exception ex) {
+            // CHECKSTYLE:ON
+            // do nothing
```
Comment (menghaoranss) [inline shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java:153]: The exception need be printed by @slf4j, and remove // do nothing

**Review by menghaoranss — CHANGES_REQUESTED** [review]

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java (hunk 1 above)
Comment (totalo) [reply shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java:142]: ok ~

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java [hunk 3]
```diff
@@ -129,23 +131,37 @@ private Type getEventChangedType(final WatchEvent event) {
...
-        // TODO
-        return false;
+        try {
+            long leaseId = client.getLeaseClient().grant(etcdProperties.getValue(EtcdPropertyKey.TIME_TO_LIVE_SECONDS)).get().getID();
+            client.getLockClient().lock(ByteSequence.from(key, StandardCharsets.UTF_8), leaseId).get(time, unit);
+            return true;
+            // CHECKSTYLE:OFF
+        } catch (final Exception ex) {
+            // CHECKSTYLE:ON
+            log.error("etcdRepository tryLock error, key = {}, time = {}, unit = {}", key, time, unit, ex);
```
Comment (menghaoranss) [inline shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java:144]: etcdRepository -> EtcdRepository , key = {} -> key: {}

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java [hunk 4]
```diff
@@ -129,23 +131,37 @@ private Type getEventChangedType(final WatchEvent event) {
...
+
     @Override
     public void releaseLock(final String key) {
-        // TODO
+        try {
+            client.getLockClient().unlock(ByteSequence.from(key, StandardCharsets.UTF_8)).get(etcdProperties.getValue(EtcdPropertyKey.CONNECTION_TIMEOUT_SECONDS), TimeUnit.SECONDS);
+            // CHECKSTYLE:OFF
+        } catch (final Exception ex) {
+            // CHECKSTYLE:ON
+            log.error("etcdRepository releaseLock error, key = {}", key, ex);
```
Comment (menghaoranss) [inline shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java:156]: etcdRepository -> EtcdRepository , key = {} -> key: {}

**Review by menghaoranss — CHANGES_REQUESTED** [review]

File: shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java (hunk 4 above)
Comment (totalo) [reply shardingsphere-governance/shardingsphere-governance-repository/shardingsphere-governance-repository-provider/shardingsphere-governance-repository-etcd/src/main/java/org/apache/shardingsphere/governance/repository/etcd/EtcdRepository.java]: ok i have fixed it

**Review by menghaoranss — APPROVED** [review]

**Pull request merged**